### PR TITLE
RHELPLAN-41563 - Logging - splitting an input/output config file into…

### PR DESCRIPTION
**Pull request by nhosoi**
… multiple ones that contains just one operation

Number of sub input/output operation could vary depending upon the logging_inputs/_outputs
values configured by the user. Configurations derived from the various operations are
generated separately from the static configurations in each sub input/output task with a
template, e.g., input_basics.j2 and output_elasticsearch.j2. The generated configuration
file name will be 2_digit_weight-input_or_output-type-item_name.conf, e.g.,
30-output-files-files_test.conf.

Updated following docs:
  design_docs/logging_subsystem.md
             /rsyslog_inputs_outputs_flows.md
             /rsyslog_templates.md

### CODE REVIEW

File: roles/rsyslog/tasks/inputs/basics/main.yml [hunk 1]
```diff
@@ -6,3 +6,16 @@
     __rsyslog_rules: "{{ __rsyslog_basics_rules }}"
   include_tasks:
     file: deploy.yml
+
+- name: Create basics input configuration files in /etc/rsyslog.d
+  vars:
+    __basics_input: "{{ lookup('template', 'input_basics.j2') }}"
+    __rsyslog_packages: []
+    __rsyslog_rules: [ { "name": "input-basics-{{ basics_item.name }}", "type": "input", "weight": "11",
+                         "sections": [ { "options": "{{ __basics_input }}" } ] } ]
```
Comment (richm) [inline roles/rsyslog/tasks/inputs/basics/main.yml:15]: do you need to define `__basics_input` here?  Can you do
```yaml

  vars:
    __rsyslog_packages: []
    __rsyslog_rules: [ { "name": "input-basics-{{ basics_item.name }}", "type": "input", "weight": "11",
                         "sections": [ { "options": "{{ lookup('template', 'input_basics.j2') }}" } ] } ]
```
?

File: roles/rsyslog/tasks/inputs/files/main.yml [hunk 2]
```diff
@@ -6,3 +6,16 @@
     __rsyslog_rules: "{{ __rsyslog_files_input_rules }}"
   include_tasks:
     file: deploy.yml
+
+- name: Create files input configuration files in /etc/rsyslog.d
+  vars:
+    __files_input: "{{ lookup('template', 'input_files.j2') }}"
+    __rsyslog_packages: []
+    __rsyslog_rules: [ { "name": "input-files-{{ files_item.name }}", "type": "input",
+                         "sections": [ { "options": "{{ __files_input }}" } ] } ]
```
Comment (richm) [inline roles/rsyslog/tasks/inputs/files/main.yml:15]: same question here about `__files_input` as above about `__basics_input`

File: roles/rsyslog/tasks/outputs/elasticsearch/main.yml [hunk 3]
```diff
@@ -59,3 +59,15 @@
     __rsyslog_rules: "{{ __rsyslog_elasticsearch_rules }}"
   include_tasks:
     file: deploy.yml
+
+- name: Create elasticsearch output configuration files in /etc/rsyslog.d
+  vars:
+    __elasticsearch_output: "{{ lookup('template', 'output_elasticsearch.j2') }}"
+    __rsyslog_packages: []
+    __rsyslog_rules: [ { "name": "output-elasticsearch-{{ es_item.name }}", "type": "output", "weight": "31",
+                         "sections": [ { "options": "{{ __elasticsearch_output }}" } ] } ]
```
Comment (richm) [inline roles/rsyslog/tasks/outputs/elasticsearch/main.yml:68]: see above - do you need to define `__elasticsearch_output` here?

File: roles/rsyslog/tasks/outputs/files/main.yml [hunk 4]
```diff
@@ -6,3 +6,15 @@
     __rsyslog_rules: "{{ __rsyslog_files_output_rules }}"
   include_tasks:
     file: deploy.yml
+
+- name: Create files output configuration files in /etc/rsyslog.d
+  vars:
+    __files_output: "{{ lookup('template', 'output_files.j2') }}"
+    __rsyslog_packages: []
+    __rsyslog_rules: [ { "name": "output-files-{{ files_item.name }}", "type": "output",
+                         "sections": [ { "options": "{{ __files_output }}" } ] } ]
```
Comment (richm) [inline roles/rsyslog/tasks/outputs/files/main.yml:15]: see above

File: roles/rsyslog/tasks/outputs/forwards/main.yml [hunk 5]
```diff
@@ -6,3 +6,15 @@
     __rsyslog_rules: "{{ __rsyslog_forwards_output_rules }}"
   include_tasks:
     file: deploy.yml
+
+- name: Create forwards output configuration files in /etc/rsyslog.d
+  vars:
+    __forwards_output: "{{ lookup('template', 'output_forwards.j2') }}"
+    __rsyslog_packages: []
+    __rsyslog_rules: [ { "name": "output-forwards-{{ forwards_item.name }}", "type": "output",
+                         "sections": [ { "options": "{{ __forwards_output }}" } ] } ]
```
Comment (richm) [inline roles/rsyslog/tasks/outputs/forwards/main.yml:15]: see above

File: roles/rsyslog/templates/input_basics.j2 [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+if
+{%     for inputname in outdict[output.name] %}
+{%       if inputname == basics_item.name %}
+{%         if not loop.first %}
+  or
+{%         endif %}
+{%         set input = indict[inputname] %}
+{%         if input.type == "basics" %}
+  ($inputname == "imjournal")
+{%         elif input.type == "basics" %}
```
Comment (richm) [inline roles/rsyslog/templates/input_basics.j2:25]: This is the same condition as above?  should this be `input.type == "files"`?

Comment (richm) [reply roles/rsyslog/templates/input_basics.j2:25]: or, are all of the inputs here `"basics"`?

File: roles/rsyslog/templates/input_files.j2 [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+{%     for inputname in outdict[output.name] %}
+{%       if inputname == files_item.name %}
+{%         if not loop.first %}
+  or
+{%         endif %}
+{%         set input = indict[inputname] %}
+{%         if input.type == "basics" %}
+  ($inputname == "imjournal")
+{%         elif input.type == "files" %}
+  ($syslogtag == "{{ input.name }}")
```
Comment (richm) [inline roles/rsyslog/templates/input_files.j2:27]: Ah, I see.

I think we should, in a separate PR, refactor this to use a single template for this - the code in `input_files.j2` is almost identical to `input_basics.j2`.

File: roles/rsyslog/tasks/inputs/basics/main.yml (hunk 1 above)
Comment (nhosoi) [reply roles/rsyslog/tasks/inputs/basics/main.yml:15]: Nope, not any more. :)  Thanks!

**Comment by richm** [comment]
lgtm

**Comment by nhosoi** [comment]
Thanks for reviewing the pr, @richm.  Merging...

File: roles/rsyslog/templates/input_files.j2 (hunk 7 above)
Comment (nhosoi) [reply roles/rsyslog/templates/input_files.j2:27]: > Ah, I see.
> 
> I think we should, in a separate PR, refactor this to use a single template for this - the code in `input_files.j2` is almost identical to `input_basics.j2`.

@richm, I've submitted a pr for sharing one template between the 2 inputs.
https://github.com/linux-system-roles/logging/pull/111

**Pull request merged**